pkg/object/bunny.go

```diff
@@ -0,0 +1,193 @@
+	return b.endpoint
+}
+
+// Limits of the object storage.
+func (b *bunnyClient) Limits() Limits {
+	return Limits{
+		IsSupportMultipartUpload: false,
+		IsSupportUploadPartCopy:  false,
+	}
+}
+
+// Get the data for the given object specified by key.
+func (b *bunnyClient) Get(key string, off int64, limit int64) (io.ReadCloser, error) {
+	if limit == -1 {
+		limit = math.MaxInt64
+	}
+	body, err := b.client.DownloadPartial(key, off, limit+off-1)
+	if err != nil {
+		return nil, err
+	}
+	return io.NopCloser(bytes.NewReader(body)), nil
+}
+
+// Put data read from a reader to an object specified by key.
+func (b *bunnyClient) Put(key string, in io.Reader) error {
+	content, readErr := io.ReadAll(in)
+	if readErr != nil {
+		return readErr
+	}
+	return b.client.Upload(key, content, true)
+}
+
+// Delete a object.
+// Requires a conditional retry, since deleting a directory or file called foo/bar requires two different calls to the Bunny API, which JuiceFS does not do
+// Deleting a directory requires a trailing slash in the key to delete, which JuiceFS does not add to the path, leading to test case failures.
+// We implement a conditional retry here to try deleting the directory if the delete for a key of the passed name fails.
+// Deleting keys that do not exist are expected to not throw an error
+func (b *bunnyClient) Delete(key string) error {
+	if err := b.client.Delete(key, false); err != nil {
+		if err.Error() == "Not Found" {
+			// Retry delete the directory with the same name
+			if errDirectoryDelete := b.client.Delete(key, true); err != nil {
+				if err.Error() == "Not Found" {
+					return nil
+				} else {
+					return errDirectoryDelete
+				}
+			}
+		}
+		return err
+	}
+	return nil
+}
+
+// ListAll returns all the objects as an channel.
+func (b *bunnyClient) ListAll(prefix string, marker string, followLink bool) (<-chan Object, error) {
+
+	out := make(chan Object)
+	go func() {
+		defer close(out)
+		b.walkObjects(prefix, marker, out)
+	}()
+
+	return out, nil
+}
+
+func (b *bunnyClient) walkObjects(prefix string, marker string, out chan<- Object) {
+	objects, err := b.client.List(prefix)
+	if err != nil {
+		return
+	}
+
+	objectsToProcess := objects
+
+	// If no objects are returned list the parent directory and continue
+	if len(objectsToProcess) == 0 {
+		objectsToProcess, err = b.client.List(path.Dir(prefix))
+		if err != nil {
+			return
+		}
+	}
+
+	markerEncountered := marker == ""
+
+	for i := 0; i < len(objectsToProcess); i++ {
+		o := objectsToProcess[i]
+		normalizedPath := normalizedObjectNameWithinZone(o)
+		if strings.HasPrefix(normalizedPath, path.Join(prefix)) {
+			if normalizedPath == marker {
+				markerEncountered = true
+			}
+			if markerEncountered {
+				if o.IsDirectory {
+					objects, err := b.client.List(normalizedPath)
+					if err != nil {
+						return
+					}
+					objectsToProcess = append(objectsToProcess, objects...)
+				}
+				out <- parseObjectMetadata(o)
```

Comment:
The objects should be ordered by key, for example,

a/
a/a
a/b
b/
b/a
b/b

Reply:
Ok, I am going to implement the list function instead of the listAll method. That should fix this behavior.